java/ql/src/semmle/code/java/dataflow/internal/rangeanalysis/SignAnalysisCommon.qll

```diff
@@ -0,0 +1,293 @@
+  or
+  lowerBound(bound, v, pos, false) and TZero() = exprSign(bound)
+  or
+  upperBound(bound, v, pos, _) and TPos() = exprSign(bound)
+  or
+  upperBound(bound, v, pos, false) and TZero() = exprSign(bound)
+  or
+  eqBound(bound, v, pos, true) and TZero() = exprSign(bound)
+  or
+  eqBound(bound, v, pos, false) and TZero() != exprSign(bound)
+}
+
+/**
+ * Holds if there is a bound that might restrict whether `v` has the sign `s`
+ * at `pos`.
+ */
+private predicate hasGuard(SsaVariable v, SsaReadPosition pos, Sign s) {
+  s = TPos() and posBound(_, v, pos)
+  or
+  s = TNeg() and negBound(_, v, pos)
+  or
+  s = TZero() and zeroBound(_, v, pos)
+}
+
+pragma[noinline]
+private Sign guardedSsaSign(SsaVariable v, SsaReadPosition pos) {
+  // SSA variable can have sign `result`
+  result = ssaDefSign(v) and
+  pos.hasReadOfVar(v) and
+  // there are guards at this position on `v` that might restrict it to be sign `result`.
+  // (So we need to check if they are satisfied)
+  hasGuard(v, pos, result)
+}
+
+pragma[noinline]
+private Sign unguardedSsaSign(SsaVariable v, SsaReadPosition pos) {
+  // SSA variable can have sign `result`
+  result = ssaDefSign(v) and
+  pos.hasReadOfVar(v) and
+  // there's no guard at this position on `v` that might restrict it to be sign `result`.
+  not hasGuard(v, pos, result)
+}
+
+/**
+ * Gets the sign of `v` at read position `pos`, when there's at least one guard
+ * on `v` at position `pos`. Each bound corresponding to a given sign must be met
+ * in order for `v` to be of that sign.
+ */
+private Sign guardedSsaSignOk(SsaVariable v, SsaReadPosition pos) {
+  result = TPos() and
+  forex(Expr bound | posBound(bound, v, pos) | posBoundOk(bound, v, pos))
+  or
+  result = TNeg() and
+  forex(Expr bound | negBound(bound, v, pos) | negBoundOk(bound, v, pos))
+  or
+  result = TZero() and
+  forex(Expr bound | zeroBound(bound, v, pos) | zeroBoundOk(bound, v, pos))
+}
+
+/** Gets a possible sign for `v` at `pos`. */
+Sign ssaSign(SsaVariable v, SsaReadPosition pos) {
+  result = unguardedSsaSign(v, pos)
+  or
+  result = guardedSsaSign(v, pos) and
+  result = guardedSsaSignOk(v, pos)
+}
+
+/** Gets a possible sign for `v`. */
+pragma[nomagic]
+Sign ssaDefSign(SsaVariable v) {
+  result = explicitSsaDefSign(v)
+  or
+  result = implicitSsaDefSign(v)
+  or
+  exists(SsaPhiNode phi, SsaVariable inp, SsaReadPositionPhiInputEdge edge |
+    v = phi and
+    edge.phiInput(phi, inp) and
+    result = ssaSign(inp, edge)
+  )
+}
+
+/** Gets a possible sign for `e`. */
+cached
+Sign exprSign(Expr e) {
+  result = certainExprSign(e)
+  or
+  not exists(certainExprSign(e)) and
+  (
+    unknownSign(e)
+    or
+    exists(SsaVariable v | getARead(v) = e | result = ssaVariableSign(v, e))
+    or
+    e =
+      any(VarAccess access |
+        not exists(SsaVariable v | getARead(v) = access) and
+        (
+          result = fieldSign(getField(access.(FieldAccess))) or
+          not access instanceof FieldAccess
```

Comment:
While we're touching this library, I think it could be nice to explicitly bind `result` also in the cases where sign is unrestricted. This just requires a dummy predicate like `predicate anySign(Sign s) { any() }` or we could write `result = any(Sign s)`, but I think I slightly prefer `anySign(result)`.